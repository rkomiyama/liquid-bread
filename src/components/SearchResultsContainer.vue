<template>
  <SearchResults
    :results="results"
    :loading="loading"
    :search="search"
    :headers="headers"
    :rowsPerPageItems="rowsPerPageItems"
  />
</template>

<script>
import axios from "axios";
import SearchResults from "./SearchResults";

/**
 * Container component that handles search results data
 */
export default {
  name: "SearchResultsContainer",
  data() {
    return {
      loading: false,
      rowsPerPageItems: [10, 20, 40, 80],
      results: []
    };
  },
  props: {
    /**
     * Quick-search string inputted by the user
     */
    search: {
      type: String
    },
    /**
     * Headers for the search results table columns
     */
    headers: {
      type: Array
    },
    /**
     * Filters that are applied to the search
     */
    filters: {
      type: Object
    }
  },
  methods: {
    /**
     * Getter function for searching beer data
     * @param {Object} params Parameters for the search
     * @public
     */
    async getResults(params = {}) {
      this.loading = true;
      this.results = await axios
        .get("https://api.punkapi.com/v2/beers", {
          params
        })
        .then(res => {
          this.loading = false;
          return res.data;
        })
        .catch(err => {
          this.loading = false;
        });
    }
  },
  watch: {
    filters(filters) {
      let searchParams = {};

      if (filters.abvGt) {
        searchParams.abv_gt = filters.abvGt;
      }
      if (filters.abvLt) {
        searchParams.abv_lt = filters.abvLt;
      }
      if (filters.ibuGt) {
        searchParams.ibu_gt = filters.ibuGt;
      }
      if (filters.ibuLt) {
        searchParams.ibu_lt = filters.ibuLt;
      }
      if (filters.ebcGt) {
        searchParams.ebc_gt = filters.ebcGt;
      }
      if (filters.ebcLt) {
        searchParams.ebc_lt = filters.ebcLt;
      }
      if (filters.beer_name) {
        searchParams.beer_name = filters.beerName.replace(/ /g, "_");
      }
      if (filters.yeast) {
        searchParams.yeast = filters.yeast.replace(/ /g, "_");
      }
      if (filters.beforeDate) {
        searchParams.brewed_before = filters.beforeDate
          .split("-")
          .reverse()
          .join("/");
      }
      if (filters.afterDate) {
        searchParams.brewed_after = filters.afterDate
          .split("-")
          .reverse()
          .join("/");
      }
      if (filters.hops) {
        searchParams.hops = filters.hops.replace(/ /g, "_");
      }
      if (filters.malt) {
        searchParams.malt = filters.malt.replace(/ /g, "_");
      }
      if (filters.food) {
        searchParams.food = filters.food.replace(/ /g, "_");
      }
      if (filters.ids) {
        searchParams.ids = filters.ids.replace(/ /g, "|");
      }

      this.getResults(searchParams);
    }
  },
  components: {
    SearchResults
  }
};
</script>

<style lang="scss" scoped>
</style>

<docs>
```html
<SearchResultsContainer :search="search" :headers="headers" :filters="filters"/>
```
</docs>