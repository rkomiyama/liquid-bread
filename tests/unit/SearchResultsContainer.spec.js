import Vue from "vue";
import Vuetify from "vuetify";
import { shallowMount } from "@vue/test-utils";
import { createRenderer } from "vue-server-renderer";
import moxios from "moxios";
import SearchResultsContainer from "@/components/SearchResultsContainer.vue";
import searchHeaders from "#/mockData/searchColumnHeaders.json";
import searchFilters from "#/mockData/searchFilters.json";
import searchResults from "#/mockData/searchResults.json";

Vue.use(Vuetify);

let wrapper;
const getResults = jest.fn();

describe("SearchResultsContainer.vue", () => {
  beforeEach(() => {
    moxios.install();
    wrapper = shallowMount(SearchResultsContainer, {
      propsData: {
        headers: searchHeaders,
        filters: searchFilters
      },
      methods: {
        getResults
      }
    });
  });

  afterEach(() => {
    moxios.uninstall();
  });

  it("calls getResults", () => {
    expect(getResults).not.toHaveBeenCalled();
    wrapper.setProps({ filters: {} });
    expect(getResults).toHaveBeenCalled();
    let newSearchFilters = JSON.parse(JSON.stringify(searchFilters));
    newSearchFilters.abvGt = "5";
    newSearchFilters.abvLt = "10";
    newSearchFilters.ibuGt = "5";
    newSearchFilters.ibuLt = "10";
    newSearchFilters.ebcGt = "5";
    newSearchFilters.ebcLt = "10";
    newSearchFilters.beer_name = "beer name";
    newSearchFilters.yeast = "yeast";
    newSearchFilters.beforeDate = "2019-03";
    newSearchFilters.afterDate = "2000-01";
    newSearchFilters.hops = "hops";
    newSearchFilters.malt = "malt";
    newSearchFilters.food = "food";
    newSearchFilters.ids = "ids";
    wrapper.setProps({ filters: newSearchFilters });
    expect(wrapper.vm.filters.abvLt).toBe("10");
    expect(getResults).toHaveBeenCalledTimes(2);
  });

  it("calls beer API successfully", done => {
    wrapper = shallowMount(SearchResultsContainer);
    const searchFiltersCopy = JSON.parse(JSON.stringify(searchFilters));
    wrapper.setProps({ filters: searchFiltersCopy });
    moxios.wait(function() {
      let request = moxios.requests.mostRecent();
      request
        .respondWith({
          status: 200,
          response: searchResults
        })
        .then(function(res) {
          expect(res.data).toEqual(searchResults);
          done();
        });
    });
  });

  it("calls beer API with an error", done => {
    wrapper = shallowMount(SearchResultsContainer);
    expect(wrapper.vm.results).toEqual([]);
    const searchFiltersCopy = JSON.parse(JSON.stringify(searchFilters));
    searchFiltersCopy.abvGt = "not a number";
    wrapper.setProps({ filters: searchFiltersCopy });
    moxios.wait(function() {
      let request = moxios.requests.mostRecent();
      request
        .respondWith({
          status: 400,
          response: "Error: Could not complete search"
        })
        .then(() => {
          expect(wrapper.vm.results).toEqual(undefined);
          done();
        });
    });
  });

  it("SearchResultsContainer has same HTML structure", () => {
    const renderer = createRenderer();
    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) throw new Error(err);
      expect(str).toMatchSnapshot();
    });
  });
});
