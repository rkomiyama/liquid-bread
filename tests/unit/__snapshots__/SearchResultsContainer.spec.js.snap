// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResultsContainer.vue SearchResultsContainer has same HTML structure 1`] = `<SearchResults-stub headers="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" results="" rowsPerPageItems="10,20,40,80"></SearchResults-stub>`;
